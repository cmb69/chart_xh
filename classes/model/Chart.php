<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Chart_XH.
 *
 * Chart_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chart_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chart_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Chart\Model;

use DOMDocument;
use DOMElement;
use DOMNode;
use Plib\Document2;
use Plib\DocumentStore2;

final class Chart implements Document2
{
    private string $name;
    private string $caption;
    /** @var list<string> */
    private array $labels = [];
    /** @var list<Dataset> */
    private array $datasets = [];

    public static function new(string $key): self
    {
        return new self(basename($key, ".xml"), "");
    }

    public static function fromString(string $contents, string $key): ?self
    {
        if ($contents === "") {
            return null;
        }
        $doc = new DOMDocument("1.0", "UTF-8");
        if (!$doc->loadXML($contents)) {
            return null;
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return null;
        }
        assert($doc->documentElement instanceof DOMElement);
        $chart = $doc->documentElement;
        $that = new self(basename($key, ".xml"), $chart->getAttribute("caption"));
        foreach ($chart->childNodes as $childNode) {
            assert($childNode instanceof DOMNode);
            if ($childNode->nodeName === "label") {
                assert($childNode instanceof DOMElement);
                $that->labels[] = (string) $childNode->nodeValue;
            } elseif ($childNode->nodeName === "dataset") {
                assert($childNode instanceof DOMElement);
                $that->datasets[] = Dataset::fromXml($childNode);
            }
        }
        return $that;
    }

    public static function create(string $name, DocumentStore2 $store): self
    {
        $that = $store->create("$name.xml", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function read(string $name, DocumentStore2 $store): ?self
    {
        return $store->read("$name.xml", self::class);
    }

    public static function update(string $name, DocumentStore2 $store): ?self
    {
        return $store->update("$name.xml", self::class);
    }

    public function __construct(string $name, string $caption)
    {
        $this->name = $name;
        $this->caption = $caption;
    }

    public function name(): string
    {
        return $this->name;
    }

    public function caption(): string
    {

        return $this->caption;
    }

    /** @return list<string> */
    public function labels(): array
    {
        return $this->labels;
    }

    /** @return list<Dataset> */
    public function datasets(): array
    {
        return $this->datasets;
    }

    public function setCaption(string $caption): void
    {
        $this->caption = $caption;
    }

    public function addLabel(string $label): void
    {
        $this->labels[] = $label;
    }

    public function addDataset(string $label, string $color): Dataset
    {
        return $this->datasets[] = new Dataset($label, $color);
    }

    public function toString(): string
    {
        $doc = new DOMDocument('1.0', 'UTF-8');
        $chart = $doc->createElement('chart');
        $chart->setAttribute("caption", $this->caption);
        $doc->appendChild($chart);
        foreach ($this->labels as $label) {
            $elt = $doc->createElement("label");
            $elt->nodeValue = $label;
            $chart->appendChild($elt);
        }
        foreach ($this->datasets as $dataset) {
            $chart->appendChild($dataset->toXml($doc));
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return "";
        }
        $doc->formatOutput = true;
        return (string) $doc->saveXML();
    }
}
