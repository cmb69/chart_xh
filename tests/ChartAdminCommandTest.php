<?php

namespace Chart;

use ApprovalTests\Approvals;
use Chart\Model\Chart;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\CsrfProtector;
use Plib\DocumentStore2 as DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class ChartAdminCommandTest extends TestCase
{
    private DocumentStore $store;
    /** @var CsrfProtector&Stub */
    private $csrfProtector;
    private View $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $chart = Chart::create("test", $this->store);
        $dataset = $chart->addDataset("one", "#ff0000");
        $dataset->addValue(1);
        $dataset->addValue(2);
        $dataset->addValue(3);
        $this->store->commit();
        $this->csrfProtector = $this->createStub(CsrfProtector::class);
        $this->csrfProtector->method("token")->willReturn("0123456789ABCDEF");
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["chart"]);
    }

    private function sut(): ChartAdminCommand
    {
        return new ChartAdminCommand("./", $this->store, $this->csrfProtector, $this->view);
    }

    public function testRendersOverview(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()($request);
        $this->assertSame("Chart – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersEditorForNewChart(): void
    {
        $request = new FakeRequest(["url" => "http://example.com/?&chart&admin=plugin_main&action=create"]);
        $response = $this->sut()($request);
        $this->assertSame("Chart – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testCreatesNewChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertFileExists(vfsStream::url("root/new.xml"));
        $this->assertSame("http://example.com/?&chart&admin=plugin_main", $response->location());
    }

    public function testCreatingIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToSaveNewChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the chart “new”!", $response->output());
    }

    public function testRendersEditorForUpdate(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
        ]);
        $response = $this->sut()($request);
        $this->assertSame("Chart – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testsReportsThatNoChartIsSelected(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You did not select a chart!", $response->output());
    }

    public function testReportsFailureToLoadChart(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=does-not-exist",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the chart “does-not-exist”!", $response->output());
    }

    public function testUpdatesChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "new chart caption",
                "labels" => "one,two,three",
                "datasets" => '[{"label":"one","color":"#ff0000","values":[1,2,3]}]'
            ],
        ]);
        $response = $this->sut()($request);
        $chart = Chart::read("test", $this->store);
        $this->assertSame("new chart caption", $chart->caption());
        $this->assertEquals(["one", "two", "three"], $chart->labels());
        $this->assertCount(1, $chart->datasets());
        $dataset = $chart->datasets()[0];
        $this->assertEquals("one", $dataset->label());
        $this->assertEquals("#ff0000", $dataset->color());
        $this->assertCount(3, $dataset->values());
        $this->assertSame("http://example.com/?&chart&admin=plugin_main&chart_name=test", $response->location());
    }

    public function testsReportsThatNoChartIsSelectedWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update",
            "post" => [
                "chart_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You did not select a chart!", $response->output());
    }

    public function testsReportsFailureToLoadChartWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=does-not-exist",
            "post" => [
                "chart_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the chart “does-not-exist”!", $response->output());
    }

    public function testUpdateIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "a new caption",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToUpdateChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "a new caption",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the chart “test”!", $response->output());
    }
}
