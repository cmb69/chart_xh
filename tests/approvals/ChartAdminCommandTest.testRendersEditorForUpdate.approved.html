
<!-- <script type="module" src=""></script>-->
<article class="chart_edit">
  <h1>Chart â€“ Edit</h1>
  <form method="post">
    <p>
      <label>
        <span>Name</span>
        <span class="chart_help">(only a-z, 0-9, and hyhpens)</span>
        <input name="name" value="test" disabled required pattern="[a-z0-9\-]+">
      </label>
    </p>
    <p>
      <label>
        <span>Caption</span>
        <textarea name="caption"></textarea>
      </label>
    </p>
    <p class="chart_labels">
      <label>
        <span>Labels</span>
        <span class="chart_help">(list of semicolon separated strings)</span>
        <textarea name="labels"></textarea>
      </label>
    </p>
    <p class="chart_datasets">
      <label>
        <span>Datasets</span>
        <textarea name="datasets">[{&quot;label&quot;:&quot;one&quot;,&quot;color&quot;:&quot;#ff0000&quot;,&quot;values&quot;:[1,2,3]}]</textarea>
      </label>
    </p>
    <p class="chart_controls">
      <button name="chart_do">Save</button>
    </p>
    <input type="hidden" name="chart_token" value="0123456789ABCDEF">
  </form>
</article>
